File: Quicksort/quicksort.py
import time
import random
import sys

def qsort(A):
    quicksort(A, 0, len(A) - 1)

def quicksort(A, p, r):
    if p < r:
        q = partition(A, p, r)
        quicksort(A, p, q - 1)
        quicksort(A, q + 1, r)

def swap(A, i, j):
    temp = A[i]
    A[i] = A[j]
    A[j] = temp

def partition(A, p, r):
    k = A[r]
    i = p - 1
    for j in range(p, r):
        if A[j] <= k:
            i += 1
            swap(A, i, j)
    swap(A, i + 1, r)
    return i + 1

File: Quicksort/Quicksort.jl
module Quicksort
    export quicksort!

    function swap!(arr::Array{T, 1}, p::Int, q::Int) where {T <: Union{Number, Char}}
        temp = arr[p]
        arr[p] = arr[q]
        arr[q] = temp
    end

    function partition(arr::Array{T, 1}, p::Int, q::Int) where {T <: Union{Number, Char}}
        # Get the pivot element
        pivot = arr[q]

        # Start the indicator of the lower elements to the left of the array
        i = p - 1

        for j ∈ p:(q - 1)
            # if the current element is smaller than the pivot
            if arr[j] < pivot
                i += 1
                swap!(arr, i, j)
            end
        end

        # Swap the pivot to the part where the lower numbers indicator just ended.
        swap!(arr, i + 1, q)
        return i + 1
    end

    function qsort!(arr::Array{T, 1}, p::Int, q::Int) where {T <: Union{Number, Char}}
        if p < q 
            mid = partition(arr, p, q)
            qsort!(arr, p, mid - 1)
            qsort!(arr, mid + 1, q)
        end
    end

    function quicksort!(arr::Array{T, 1}) where {T <: Union{Number, Char}}
        qsort!(arr, 1, length(arr))
    end
end